starknet_py/net/gateway_client.py

```diff
@@ -0,0 +1,328 @@
+        return TransactionSchema().load(res["transaction"], unknown=EXCLUDE)
+
+    async def get_block(
+        self,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> StarknetBlock:
+        block_identifier = self._get_block_identifier(
+            block_hash=block_hash, block_number=block_number
+        )
+
+        res = await self._feeder_gateway_client.call(
+            method_name="get_block", params=block_identifier
+        )
+        return StarknetBlockSchema().load(res, unknown=EXCLUDE)
+
+    async def get_state_update(
+        self,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> BlockState:
+        # TODO implement
+        pass
+
+    async def get_storage_at(
+        self,
+        contract_address: Union[int, str],
+        key: int,
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> int:
+        block_identifier = self._get_block_identifier(
+            block_hash=block_hash, block_number=block_number
+        )
+
+        res = await self._feeder_gateway_client.call(
+            method_name="get_storage_at",
+            params={
+                **{
+                    "contractAddress": str(hex(contract_address))
+                    if isinstance(contract_address, int)
+                    else contract_address,
+                    "key": key,
+                },
+                **block_identifier,
+            },
+        )
+        res = typing.cast(str, res)
+        return int(res, 16)
+
+    async def get_transaction_receipt(
+        self, tx_hash: Union[int, str]
+    ) -> TransactionReceipt:
+        print("hash " + str(tx_hash))
+        res = await self._feeder_gateway_client.call(
+            method_name="get_transaction_receipt",
+            params={
+                "transactionHash": str(hex(tx_hash))
+                if isinstance(tx_hash, int)
+                else tx_hash
+            },
+        )
+        print(res)
+        return TransactionReceiptSchema().load(res, unknown=EXCLUDE)
+
+    async def get_code(
+        self,
+        contract_address: Union[int, str],
+        block_hash: Optional[Union[int, str]] = None,
+        block_number: Optional[int] = None,
+    ) -> ContractCode:
+        block_identifier = GatewayClient._get_block_identifier(
+            block_hash=block_hash, block_number=block_number
+        )
+        params = {
+            **{
+                "contractAddress": str(hex(contract_address))
+                if isinstance(contract_address, int)
+                else contract_address
+            },
+            **block_identifier,
+        }
+
+        res = await self._feeder_gateway_client.call(
+            method_name="get_code", params=params
+        )
+
+        if len(res["bytecode"]) == 0:
+            raise NoContractFoundError(
```

Comment:
Is defining errors like this in separate file and then having them in docstrings for`BaseClient` a good idea>

Reply:
I don't see docstrings here. But yes, documenting them and exposing them is a good idea :)

Reply:
I've pushed an first version of custom errors in `client_errors.py`. There's one thing to consider though, RPC and gateway throw completely different errors so I think it'd be most resonable to wrap _most of these_ in uniform `ClientError` and let the user handle them.

Instead we'd have to write at least a dozen of custom exceptions and then handle them.

Reply:
🤔  shouldn't we map them?